Add Get/SetOptimalCost in MathematicalProgram

EricCousineau-TRI opened the pull request:
To address #5647, this updates the `MathematicalProgram` interface to expose the aggregate cost at the optimal solution, updates each solver implementation to use this interface, and updates existing solver tests to check this behavior.

Further actions regarding how the decision variables and cost will synchronize with failures will be addressed in #5681.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5685)
<!-- Reviewable:end -->

hongkai-dai commented:
Could you also add tests on these optimal cost? One place to add the test is in `quadratic_program_examples.h`. For each `CheckSolution` function, we can get the optimal cost, and then compute the cost by hand, then compare the results. Similarly in `linear_program_examples.h` and `optimization_examples.h`.

BTW, when you add the reviewers, I think it is better to do it in reviewable. You can add a comment like "+@hongkai-dai for feature review", and then reviewable will notify me to review, and it also shows up on github. When you add people to review on github, it only shows up on github, not in reviewable.

+@hongkai-dai for feature review.

---

Reviewed 8 of 15 files at r1.
Review status: 8 of 15 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-:-Kgagbx15IQK7SrcBxSl:berovw1)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Yup, I will add reviews via Reviewable in the future. Thank you for the pointer!

I believe I had added some of those tests, but for some reason Reviewable started off showing the diff between r1 and r1 for the test files, and I think it may have done the same for your view.

Can you double-check which revisions it's checking, and if it's only showing those?

---

Review status: 8 of 15 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-:-KgaiyT8R9IcNGSC_FDC:bfp0e32)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Yes, I see the change on `linear_program_examples` and `optimization_examples`, but I do not see the change on `quadratic_program_examples`, could you double check if you have included that in your change?

---

Reviewed 7 of 15 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.cc, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgaidD6mZXGf8pc_SIZ:-KgaidD6mZXGf8pc_SI_:bottr81) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L6)):*
> ```cpp
> 
> #include "drake/common/eigen_matrix_compare.h"
> #include "drake/math/autodiff.h"
> ```

Do not need to include this?

---

*[drake/solvers/test/optimization_examples.cc, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgaiEbD2NP6EJYYov0q:-KgaiEbD2NP6EJYYov0r:b-o05onv) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L53)):*
> ```cpp
>     }
> 
>     binding.constraint()->Eval(math::initializeAutoDiff(this_x), ty);
> ```

I do not think you need to use autodiff? Actually you may use the function `EvalBindingAtSolution()`, so this for loop can be replaced as
```C++
for (const auto& binding : prog.GetAllCosts()) {
  cost += prog.EvalBindingAtSolution(binding)(0);
}
```

---

*[drake/solvers/test/optimization_examples.cc, line 474 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgajUlDbWG0iQ_pMx1H:-KgajUlDbWG0iQ_pMx1I:b-6xxw5q) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L474)):*
> ```cpp
>   //    prog_->SetValue(expected_full, x_, expected_)
>   //    prog_->SetValue(expected_full, y_, expected_)
>   Eigen::VectorXd expected_full(6);
> ```

Use `Eigen::Matrix<double, 6, 1>` instead of `Eigen::VectorXd`. We prefer static sized object, rather than dynamic sized.

---

*[drake/solvers/test/optimization_examples.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-Kgaj-W7zzEJIxbmpcly:-Kgaj-W7zzEJIxbmpclz:b-e97njd) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.h#L30)):*
> ```objc
> };
> 
> double evaluate_cost(const MathematicalProgram& prog,
> ```

You do not need to pass `x` there. `MathematicalProgram` stores the solution, and we can call `MathematicalProgram::EvalBindingAtSolution()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-:-Kgajrqo_XZJgasBD3vo:b-o83l5)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[drake/solvers/test/equality_constrained_qp_solver_test.cc, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamEMHR07C4Qtvps48:-KgamEMHR07C4Qtvps49:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/equality_constrained_qp_solver_test.cc#L72)):*
> ```cpp
>             << "\tExpected: " << expected_answer.transpose()
>             << "\tActual: " << actual_answer.transpose();
>   expected_optimal_cost = 0;
> ```

`const double`

---

*[drake/solvers/test/equality_constrained_qp_solver_test.cc, line 110 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamBxV0Dwg1BLxhi77:-KgamBxV0Dwg1BLxhi78:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/equality_constrained_qp_solver_test.cc#L110)):*
> ```cpp
>   EXPECT_TRUE(CompareMatrices(expected_answer, x_value, 1e-10,
>                               MatrixCompareType::absolute));
>   double expected_optimal_cost = -0.25;
> ```

`const double`

---

*[drake/solvers/test/linear_program_examples.cc, line 135 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-Kgam4v8DucqcZXnVdM4:-Kgam4v8DucqcZXnVdM5:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/linear_program_examples.cc#L135)):*
> ```cpp
>   EXPECT_TRUE(CompareMatrices(prog()->GetSolution(x_), x_expected_, tol,
>                               MatrixCompareType::absolute));
>   double optimal_cost_expected = evaluate_cost(*prog(), x_expected_);
> ```

`const double`

---

*[drake/solvers/test/linear_program_examples.cc, line 176 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-Kgam3NZg5kVlX8DafPg:-Kgam3NZg5kVlX8DafPh:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/linear_program_examples.cc#L176)):*
> ```cpp
>   EXPECT_TRUE(CompareMatrices(prog()->GetSolution(x_), x_expected_, tol,
>                               MatrixCompareType::absolute));
>   double optimal_cost_expected = evaluate_cost(*prog(), x_expected_);
> ```

`const double`

---

*[drake/solvers/test/linear_program_examples.cc, line 264 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-Kgam-J4CZobsslshtfE:-Kgam-J4CZobsslshtfF:bnknxht) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/linear_program_examples.cc#L264)):*
> ```cpp
>   EXPECT_TRUE(CompareMatrices(prog()->GetSolution(x_), x_expected_, tol,
>                               MatrixCompareType::absolute));
>   double optimal_cost_expected = evaluate_cost(*prog(), x_expected_);
> ```

use `const double` rather than `double`?

---

*[drake/solvers/test/optimization_examples.cc, line 230 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamQ0Np_CpT3itU-XA:-KgamQ0Np_CpT3itU-XB:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L230)):*
> ```cpp
> bool NonConvexQPproblem1::CheckSolution() const {
>   const auto& x_value = prog_->GetSolution(x_);
>   double optimal_cost_actual = prog_->GetOptimalCost();
> ```

`const double`

---

*[drake/solvers/test/optimization_examples.cc, line 300 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamOS8ru3WP02KqSyM:-KgamOS8ru3WP02KqSyN:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L300)):*
> ```cpp
>   const auto& x_value = prog_->GetSolution(x_);
>   double optimal_cost_actual = prog_->GetOptimalCost();
>   double optimal_cost_expected = evaluate_cost(*prog_, x_expected_);
> ```

`const double`

---

*[drake/solvers/test/optimization_examples.cc, line 370 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamMzjld--VdRjncdR:-KgamMzjld--VdRjncdS:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L370)):*
> ```cpp
> bool LowerBoundedProblem::CheckSolution() const {
>   const auto& x_value = prog_->GetSolution(x_);
>   double optimal_cost_actual = prog_->GetOptimalCost();
> ```

`const double`

---

*[drake/solvers/test/optimization_examples.cc, line 476 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamLGyCD83MrG3EU6a:-KgamLGyCD83MrG3EU6b:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L476)):*
> ```cpp
>   Eigen::VectorXd expected_full(6);
>   expected_full << expected_, expected_;
>   double optimal_cost_actual = prog_->GetOptimalCost();
> ```

`const double`

---

*[drake/solvers/test/optimization_examples.cc, line 644 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamJrLo5OUA4Z7CwnQ:-KgamJrLo5OUA4Z7CwnR:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L644)):*
> ```cpp
>     auto t_rotated_lorentz_value =
>         prog_rotated_lorentz_->GetSolution(t_rotated_lorentz_);
>     double optimal_cost_actual = prog_lorentz_->GetOptimalCost();
> ```

`const double`

---

*[drake/solvers/test/optimization_examples.cc, line 655 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamIV0aNo-U-otuk1Z:-KgamIV0aNo-U-otuk1_:bo7jn4n) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L655)):*
> ```cpp
>     auto x_lorentz_value = prog_lorentz_->GetSolution(x_lorentz_);
>     auto t_lorentz_value = prog_lorentz_->GetSolution(t_lorentz_);
>     double optimal_cost_actual = prog_lorentz_->GetOptimalCost();
> ```

`const double`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[drake/solvers/equality_constrained_qp_solver.cc, line 113 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgbR0cblgnUHiLyU8MB:-KgbR0cblgnUHiLyU8MC:b-h9ejg6) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/equality_constrained_qp_solver.cc#L113)):*
> ```cpp
> 
>     // Solve G*x = A'y - c
>     Eigen::VectorXd x = llt.solve(A.transpose() * lambda - c);
> ```

BTW, consider marking this `const`:

https://google.github.io/styleguide/cppguide.html#Use_of_const

Ditto regarding `optimal_cost` below.

Ditto for all other instances in this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
I had forgotten the `quadratic_program_examples`, and have now included this - thanks for catching that! I have also updated and simplified the testing code.

Ran a quick local qualification:
```
$ cd drake/solvers
$ bazel test :equality_constrained_qp_solver_test :mathematical_program_test :moby_lcp_solver_test :ipopt_solver_test :linear_system_solver_test :nonlinear_program_test :snopt_solver_test
```

---

Review status: 10 of 16 files reviewed at latest revision, 16 unresolved discussions.

---

*[drake/solvers/equality_constrained_qp_solver.cc, line 113 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgbR0cblgnUHiLyU8MB:-KgoBbAm26dqX_1qIs1S:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/equality_constrained_qp_solver.cc#L113)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider marking this `const`:

https://google.github.io/styleguide/cppguide.html#Use_of_const

Ditto regarding `optimal_cost` below.

Ditto for all other instances in this PR.
</blockquote></details>

Done.

---

*[drake/solvers/test/equality_constrained_qp_solver_test.cc, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamEMHR07C4Qtvps48:-KgoBgXv0pnrgOtdim3W:b-rb8yr8) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/equality_constrained_qp_solver_test.cc#L72)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done, inlined the expressions to minimize code change.

---

*[drake/solvers/test/equality_constrained_qp_solver_test.cc, line 110 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamBxV0Dwg1BLxhi77:-KgoBn5s8UEU9b-0_kHu:b-3sac2k) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/equality_constrained_qp_solver_test.cc#L110)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done, as above.

---

*[drake/solvers/test/linear_program_examples.cc, line 135 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-Kgam4v8DucqcZXnVdM4:-KgoBr6EKRE84n62ISBY:b-3sac2k) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/linear_program_examples.cc#L135)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done, as above.

---

*[drake/solvers/test/linear_program_examples.cc, line 176 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-Kgam3NZg5kVlX8DafPg:-KgoBshbfx4DGya7nQLr:b-3sac2k) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/linear_program_examples.cc#L176)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done, as above.

---

*[drake/solvers/test/linear_program_examples.cc, line 264 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-Kgam-J4CZobsslshtfE:-KgoBuGlolRI0CzfQIyz:b-3sac2k) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/linear_program_examples.cc#L264)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

use `const double` rather than `double`?
</blockquote></details>

Done, as above.

---

*[drake/solvers/test/optimization_examples.cc, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgaidD6mZXGf8pc_SIZ:-KgoBwTLTZ-2cCMn4EDv:b-tflyyd) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L6)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Do not need to include this?
</blockquote></details>

Done. Removed according to suggestion below.

---

*[drake/solvers/test/optimization_examples.cc, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgaiEbD2NP6EJYYov0q:-KgoC0Klqu79taXQsIy3:b-y3q8gk) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L53)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I do not think you need to use autodiff? Actually you may use the function `EvalBindingAtSolution()`, so this for loop can be replaced as
```C++
for (const auto& binding : prog.GetAllCosts()) {
  cost += prog.EvalBindingAtSolution(binding)(0);
}
```
</blockquote></details>

Done. Thank you for you suggestion!
I also changed the scope of the checking to ensure that the cost re-evaluated at the solution matches the cost stored by the `MathematicalProgram`.

---

*[drake/solvers/test/optimization_examples.cc, line 230 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamQ0Np_CpT3itU-XA:-KgoC379ppRxN0Cr9Q7p:b-3sac2k) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L230)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done, as above.

---

*[drake/solvers/test/optimization_examples.cc, line 300 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamOS8ru3WP02KqSyM:-KgoC5cJTMh9ahugmkDm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L300)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.cc, line 370 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamMzjld--VdRjncdR:-KgoC88Q2TVzJeKJiziH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L370)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.cc, line 474 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgajUlDbWG0iQ_pMx1H:-KgoCWcZtMvbCDsk0ppy:b-a9wca5) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L474)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Use `Eigen::Matrix<double, 6, 1>` instead of `Eigen::VectorXd`. We prefer static sized object, rather than dynamic sized.
</blockquote></details>

Done. Simplified to ensure stored cost is accurate to re-evaluated cost, as mentioned above.

---

*[drake/solvers/test/optimization_examples.cc, line 476 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamLGyCD83MrG3EU6a:-KgoCcde7uM-hDRwBNwS:b-3sac2k) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L476)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done, as above.

---

*[drake/solvers/test/optimization_examples.cc, line 644 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamJrLo5OUA4Z7CwnQ:-KgoCdv3N-5acXETavch:b-3sac2k) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L644)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done, as above.

---

*[drake/solvers/test/optimization_examples.cc, line 655 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgamIV0aNo-U-otuk1Z:-KgoCfAl2QYn3m0A02Jq:b-3sac2k) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.cc#L655)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const double`
</blockquote></details>

Done, as above.

---

*[drake/solvers/test/optimization_examples.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-Kgaj-W7zzEJIxbmpcly:-Kgc9FDN1FTuejk-koYP:bynfohy) ([raw file](https://github.com/robotlocomotion/drake/blob/f4ada0ab266f8536a37228850f833f5c5c652b93/drake/solvers/test/optimization_examples.h#L30)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

You do not need to pass `x` there. `MathematicalProgram` stores the solution, and we can call `MathematicalProgram::EvalBindingAtSolution()`.
</blockquote></details>

Done. As mentioned above, I've changed the test checking to generally ensure that the stored cost is the same as the cost re-evaluated at the actual solution.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-:-KgoD1CWNR4d76zy7grA:b-ju6mg)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 6 files at r2.
Review status: 11 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.cc, line 39 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgoSIABG47R__AD7ISO:-KgoSIABG47R__AD7ISP:b3zlgu7) ([raw file](https://github.com/robotlocomotion/drake/blob/af279d03a35561498779f110a0fe8a98495ac2e9/drake/solvers/test/optimization_examples.cc#L39)):*
> ```cpp
> 
> double EvaluateSolutionCost(const MathematicalProgram &prog) {
>   Eigen::VectorXd cost = Eigen::VectorXd::Zero(1);
> ```

nit: You may do
```C++ 
double cost{0};
for (const auto& binding : prog.GetAllCosts()) {
  cost += prog.EvalBindingAtSolution(binding)(0);
}
return cost;
```
So as to avoid creating `cost` as a `Eigen::VectorXd` object.

---

*[drake/solvers/test/optimization_examples.cc, line 50 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgoT3bCtOAT5hUk0bWh:-KgoT3bCtOAT5hUk0bWi:bzgcbz5) ([raw file](https://github.com/robotlocomotion/drake/blob/af279d03a35561498779f110a0fe8a98495ac2e9/drake/solvers/test/optimization_examples.cc#L50)):*
> ```cpp
>  * be nearly the same as the cost reevaluated at the solution
>  */
> void ExpectSolutionCostAccurate(const MathematicalProgram &prog, double tol) {
> ```

Maybe it is better to just keep either `ExpectSolutionCostAccurate` or `CheckSolutionCostAccurate`? Duplication can lead to confusion. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 6 files at r2.
Review status: 12 of 16 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 4 of 6 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
*[drake/solvers/test/optimization_examples.cc, line 50 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgoT3bCtOAT5hUk0bWh:-KgoVrn6PsAsToNYquoO:b5ejdtw) ([raw file](https://github.com/robotlocomotion/drake/blob/af279d03a35561498779f110a0fe8a98495ac2e9/drake/solvers/test/optimization_examples.cc#L50)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Maybe it is better to just keep either `ExpectSolutionCostAccurate` or `CheckSolutionCostAccurate`? Duplication can lead to confusion. 
</blockquote></details>

The only issue here is that some of the existing `CheckSolution` methods return a boolean, whereas others `CheckSolution` methods use unittest `EXPECT_*` methods, and they wouldn't mix as well given that `EXPECT_TRUE(CheckSolutionCostAccurate(...))` would not provide as much information, and changing the `bool CheckSolution` methods to use `ExpectSolutionCostAccurate` would be inconsistent.

Which solution do you prefer?

1. Refactor the `bool CheckSolution` methods to mirror the `void CheckSolution` so as to use expectations (and provide more information)? (Cleanest, would take a little longer, but not much)
2. Inject expectations into the `bool CheckSolution` methods, and possibly later refactor the `CheckSolution`.
3. Use `EXPECT_TRUE(CheckSolutionCostAccurate(...))` in the expectations, at a loss of information.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.cc, line 50 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgoT3bCtOAT5hUk0bWh:-KgoYhYcCRrkzl8lupWl:b-khhtw5) ([raw file](https://github.com/robotlocomotion/drake/blob/af279d03a35561498779f110a0fe8a98495ac2e9/drake/solvers/test/optimization_examples.cc#L50)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

The only issue here is that some of the existing `CheckSolution` methods return a boolean, whereas others `CheckSolution` methods use unittest `EXPECT_*` methods, and they wouldn't mix as well given that `EXPECT_TRUE(CheckSolutionCostAccurate(...))` would not provide as much information, and changing the `bool CheckSolution` methods to use `ExpectSolutionCostAccurate` would be inconsistent.

Which solution do you prefer?

1. Refactor the `bool CheckSolution` methods to mirror the `void CheckSolution` so as to use expectations (and provide more information)? (Cleanest, would take a little longer, but not much)
2. Inject expectations into the `bool CheckSolution` methods, and possibly later refactor the `CheckSolution`.
3. Use `EXPECT_TRUE(CheckSolutionCostAccurate(...))` in the expectations, at a loss of information.
</blockquote></details>

I think using `ExpectSolutionCostAccurate` is better, as you said, `EXPECT_TRUE(CheckSolutionCostAccurate)` loses information. So maybe in the `optimization_examples.cc`, you could change the `CheckSolution` method, instead of returning a boolean, we can use `EXPECT_NEAR`, `EXPECT_EQ` inside `CheckSolution` function?

Feel free to leave this as a TODO, and file a separate PR on this issue, if you think it is outside the scope of this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 14 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.cc, line 39 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgoSIABG47R__AD7ISO:-KgpkBfa-XVPxdPBt6EC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/af279d03a35561498779f110a0fe8a98495ac2e9/drake/solvers/test/optimization_examples.cc#L39)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

nit: You may do
```C++ 
double cost{0};
for (const auto& binding : prog.GetAllCosts()) {
  cost += prog.EvalBindingAtSolution(binding)(0);
}
return cost;
```
So as to avoid creating `cost` as a `Eigen::VectorXd` object.
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.cc, line 50 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgoT3bCtOAT5hUk0bWh:-KgpkEBkJlxcfIDdUiYR:b-zfp3yv) ([raw file](https://github.com/robotlocomotion/drake/blob/af279d03a35561498779f110a0fe8a98495ac2e9/drake/solvers/test/optimization_examples.cc#L50)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think using `ExpectSolutionCostAccurate` is better, as you said, `EXPECT_TRUE(CheckSolutionCostAccurate)` loses information. So maybe in the `optimization_examples.cc`, you could change the `CheckSolution` method, instead of returning a boolean, we can use `EXPECT_NEAR`, `EXPECT_EQ` inside `CheckSolution` function?

Feel free to leave this as a TODO, and file a separate PR on this issue, if you think it is outside the scope of this PR.
</blockquote></details>

Done.

It was a simple enough change in this case. I think at some point, it would merit refactoring the testing a bit to use the same optimization problem specification (having all of them extending from `OptimizationProgram`), but we can consider crossing the bridge at a later point.
Would you like me to create a separate issue for this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 14 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/optimization_examples.cc, line 50 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgoT3bCtOAT5hUk0bWh:-Kgpl03rOJqpej_jotAc:b-o3ly3o) ([raw file](https://github.com/robotlocomotion/drake/blob/af279d03a35561498779f110a0fe8a98495ac2e9/drake/solvers/test/optimization_examples.cc#L50)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done.

It was a simple enough change in this case. I think at some point, it would merit refactoring the testing a bit to use the same optimization problem specification (having all of them extending from `OptimizationProgram`), but we can consider crossing the bridge at a later point.
Would you like me to create a separate issue for this?
</blockquote></details>

Great! Thanks for the change.

I was halfway through refactoring all the optimization tests, and then dragged away to other project. It would be great if you could create a separate issue. Currently we still have some second order conic programming tests and semidefinite programming tests left in `convex_optimization_test`, that we should refactor. We should also have them all extended from `OptimizationProgram`. And `OptimizationProgram` should just be a derived class of `MathematicalProgram`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> +@sammy-tri for platform review. 

---

Review status: 14 of 18 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-:-KgplYxmQobPtr5Jmuol:bjlm7j9)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 10 of 15 files at r1, 4 of 6 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/solvers/moby_lcp_solver.cc, line 199 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KguFqLO-x2jZRJKjdB0:-KguFqLQZ1TIZNNjRtfU:bki6cr2) ([raw file](https://github.com/robotlocomotion/drake/blob/e04043cd81c142a58b1c0f59242b04836c954ec7/drake/solvers/moby_lcp_solver.cc#L199)):*
> ```cpp
>     }
>     prog.SetDecisionVariableValues(binding.variables(), constraint_solution);
>     prog.SetOptimalCost(std::numeric_limits<double>::quiet_NaN());
> ```

Is there a reason we use NaN here (when there is no objective), but 0. in linear_system_solver.cc (which also has no objective)?

---

*[drake/solvers/test/gurobi_solver_test.cc, line 101 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KguGhStQXnVhzkwtrhx:-KguGhStQXnVhzkwtrhy:b-nz5qw6) ([raw file](https://github.com/robotlocomotion/drake/blob/e04043cd81c142a58b1c0f59242b04836c954ec7/drake/solvers/test/gurobi_solver_test.cc#L101)):*
> ```cpp
>       EXPECT_TRUE(CompareMatrices(x_value, x_expected, 1E-6,
>                                   MatrixCompareType::absolute));
>       EXPECT_TRUE(!std::isnan(prog.GetOptimalCost()));
> ```

I feel like this should have a comment for why we don't make any stronger assertions about the cost than it isn't NaN.  (perhaps why ExpectSolutionCostAccurate wouldn't come up with the right value here?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 16 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/solvers/moby_lcp_solver.cc, line 199 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KguFqLO-x2jZRJKjdB0:-KguHlmerl22h0Sw0WMF:b7v84a2) ([raw file](https://github.com/robotlocomotion/drake/blob/e04043cd81c142a58b1c0f59242b04836c954ec7/drake/solvers/moby_lcp_solver.cc#L199)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Is there a reason we use NaN here (when there is no objective), but 0. in linear_system_solver.cc (which also has no objective)?
</blockquote></details>

Changed it to 0. The original intent was to use NaN as the solver does not optimize over a cost function, but you have a point, and I discussed it with Hongkai and he mentioned that 0 would make more sense in terms of the convex quadratic minimization formulation of the LCP.

---

*[drake/solvers/test/gurobi_solver_test.cc, line 101 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KguGhStQXnVhzkwtrhx:-KguoC9bLChWOebyDaGS:by1nv9l) ([raw file](https://github.com/robotlocomotion/drake/blob/e04043cd81c142a58b1c0f59242b04836c954ec7/drake/solvers/test/gurobi_solver_test.cc#L101)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I feel like this should have a comment for why we don't make any stronger assertions about the cost than it isn't NaN.  (perhaps why ExpectSolutionCostAccurate wouldn't come up with the right value here?)
</blockquote></details>

Done. I had missed this when reimplementing with `ExpectSolutionCostAccurate`. Thanks!

---

*[drake/solvers/test/optimization_examples.cc, line 50 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-KgoT3bCtOAT5hUk0bWh:-KguqhIQja4rHpCKlaBo:buh3dzv) ([raw file](https://github.com/robotlocomotion/drake/blob/af279d03a35561498779f110a0fe8a98495ac2e9/drake/solvers/test/optimization_examples.cc#L50)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Great! Thanks for the change.

I was halfway through refactoring all the optimization tests, and then dragged away to other project. It would be great if you could create a separate issue. Currently we still have some second order conic programming tests and semidefinite programming tests left in `convex_optimization_test`, that we should refactor. We should also have them all extended from `OptimizationProgram`. And `OptimizationProgram` should just be a derived class of `MathematicalProgram`.
</blockquote></details>

Done. Added #5718.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending CI issues

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-:-Kgus9UDn8Jj7J7Hstx2:b1yjwf0)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
`cpplint` error, had not ran the `cpplint_wrapper` when re-submitting. Fixed the issue.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5685#-:-Kgv-laIT2d-3DiISoJU:b-z3yrev)*
<!-- Sent from Reviewable.io -->
